Engagements cleanup and enhancements

thomasplevy opened the pull request:
## Description

Refactors admin panel code to share image upload field HTML and JS between metaboxes and settings fields

## How has this been tested?

Manually, existing unit tests

This needs more testing, I plan to write some e2e tests (maybe)

## Screenshots <!-- if applicable -->

## Types of changes

Updates

## Checklist:
- [ ] My code has been tested.
- [x] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [x] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

Commit:
Merge pull request #1266 from gocodebox/dev

Dev to Trunk

thomasplevy committed:
secure cookie during cookie reset

thomasplevy committed:
dont use deprecated .ready()

thomasplevy committed:
fix docs

thomasplevy committed:
put var declaration inside if statement

thomasplevy committed:
don't use .ready(), deprecated in jQuery 3.0

thomasplevy committed:
update default images, per #1081

thomasplevy committed:
refactor admin upload fields to share code between settings & metabox

update cert and achievement metaboxes to use new field

update upload JS to be shared between metaboxes and settings

remove unused JS for cert and achievement metaboxs

add a singleton trait pattern

add an abstract engagement class for shared code between different engagements

thomasplevy committed:
remove unneeded code

thomasplevy committed:
remove unneeded code

thomasplevy committed:
fix double output of field description

thomasplevy committed:
use helper methods in favor of hardcoded paths

thomasplevy committed:
update docs

thomasplevy committed:
fixcs

thomasplevy commented:
@eri-trabiccolo I included the code from #1270 in this PR in order to include additional settings/css changes that may be required.

Commit:
Merge pull request #1267 from gocodebox/issue/1227

secure cookie during cookie reset

eri-trabiccolo commented on assets/js/llms-admin-settings.js:
```diff
@@ -105,74 +93,39 @@
+		 * @deprecated [version]
+		 *
+		 * @param {Object} $btn  The jQuery object for clicked button.
+		 * @param {Object} event JS event object.
```
```suggestion
		 * @param {Object} e     JS event object.
```

eri-trabiccolo commented on assets/js/llms-admin-settings.js:
```diff
@@ -105,74 +93,39 @@
+		 * @since 3.7.3
+		 * @deprecated [version]
+		 *
+		 * @param {Object}  $btn The jQuery object of the clicked button
+		 * @param {Integer} id   WP Attachment ID of the image
```
```suggestion
		 * @param {Object}  $btn The jQuery object of the clicked button.
		 * @param {Integer} id   WP Attachment ID of the image.
```

eri-trabiccolo commented on assets/js/llms-admin-settings.js:
```diff
@@ -105,74 +93,39 @@
 		 */
 		this.update_image = function( $btn, id, src ) {
+			console.log( 'llms.admin_settings.update_image() is deprecated since [version]! Use llms.admin_fields.image.update_image() instead.' );
+			return window.llms.admin_fields.image.update_image( $btn, e );
```
we don't have any `e` here

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+		this.bind = function() {
+
+			$uploads.on( 'click', function( e ) {
+				e.preventDefault();
```
```suggestion
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+			} );
+
+			$( '.llms-image-field-remove' ).on( 'click', function( e ) {
+				e.preventDefault();
```
```suggestion
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+				self.open_media_lib( $( this ), e );
+			} );
+
+			$( '.llms-image-field-remove' ).on( 'click', function( e ) {
```
```suggestion
			$( '.llms-image-field-remove' ).on( 'click', function() {
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+		 *
+		 * @return {Void}
+		 */
+		this.bind = function() {
```
We don't need to prevent the default behavior of the click event on button, so we can remove those lines, right?
Also, it doesn't seem that `open_media_lib()` uses the `event` parameter at all.

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+		 * @since [version]
+		 *
+		 * @param {Object} $btn  The jQuery object for clicked button.
+		 * @param {Object} event JS event object.
+		 * @return {Void}
+		 */
+		this.open_media_lib = function( $btn, event ) {
```
```suggestion
		 * @return {Void}
		 */
		this.open_media_lib = function( $btn ) {
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+		 */
+		this.open_media_lib = function( $btn, event ) {
+
+			var self  = this,
+				frame = null;
+
+			if ( ! frame ) {
+				var title       = $btn.attr( 'data-frame-title' ) || LLMS.l10n.translate( 'Select an Image' ),
+					button_text = $btn.attr( 'data-frame-button' ) || LLMS.l10n.translate( 'Select Image' );
+					frame       = wp.media.frames.file_frame = wp.media( {
+						title: title,
+						button: {
+							text: button_text,
+						},
+						multiple: false,
+					} );
+			}
```
This, as in other places, looks like an indentation issue when declaring vars.
It's not in our code editors, but I'm not sure if this is the _right_ way to indent, nor if there's one.

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+		this.update_image = function( $btn, id, src ) {
+
+			var $input   = $( '#' + $btn.attr( 'data-id' ) ),
+				$preview = $btn.prevAll( 'img.llms-image-field-preview' )
```
```suggestion
				$preview = $btn.prevAll( 'img.llms-image-field-preview' ),
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+		 *
+		 * @since [version]
+		 *
+		 * @param {Object}  $btn The jQuery object of the clicked button
+		 * @param {Integer} id   WP Attachment ID of the image
```
```suggestion
		 * @param {Object}  $btn The jQuery object of the clicked button.
		 * @param {Integer} id   WP Attachment ID of the image.
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+
+			if ( ! frame ) {
+				var title       = $btn.attr( 'data-frame-title' ) || LLMS.l10n.translate( 'Select an Image' ),
+					button_text = $btn.attr( 'data-frame-button' ) || LLMS.l10n.translate( 'Select Image' );
```
```suggestion
					button_text = $btn.attr( 'data-frame-button' ) || LLMS.l10n.translate( 'Select Image' ),
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+			$input.val( id );
+			$preview.attr( 'src', src );
+
+			if ( '' !== id ) {
+				$remove.removeClass( 'hidden' );
+			} else {
+				$remove.addClass( 'hidden' );
+			}
```
some golf...
```suggestion
			$remove.toggleClass( 'hidden', '' === id );
```

eri-trabiccolo commented on assets/js/llms-admin.js:
```diff
@@ -2,15 +2,124 @@
+
+			var $input   = $( '#' + $btn.attr( 'data-id' ) ),
+				$preview = $btn.prevAll( 'img.llms-image-field-preview' )
+				$remove  = $btn.hasClass( 'llms-image-field-remove' ) ? $btn : $btn.next( 'input.llms-image-field-remove' );
```
```suggestion
				$remove  = $btn.hasClass( 'llms-image-field-remove' ) ? $btn : $btn.next( 'button.llms-image-field-remove' );
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-engagement.php:
```diff
@@ -0,0 +1,109 @@
+
+		$type = $this->get_type();
+
+		$src = LLMS_PLUGIN_URL . "/assets/images/optional_{$type}.png";
```
```suggestion
		$src = LLMS_PLUGIN_URL . "assets/images/optional_{$type}.png";
```

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/fields/llms.class.meta.box.image.php:
```diff
@@ -4,68 +4,72 @@
+	 * @since Unknown
+	 *
 	 * @param array $_field Array containing information about field
+	 *
```
```suggestion
```

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/fields/llms.class.meta.box.image.php:
```diff
@@ -4,68 +4,72 @@
 	 *
+	 * @since Unknown
+	 *
 	 * @param array $_field Array containing information about field
```
```suggestion
	 * @param array $_field Array containing information about field.
```

eri-trabiccolo commented on includes/admin/post-types/meta-boxes/fields/llms.class.meta.box.textarea.tags.php:
```diff
@@ -20,15 +20,20 @@ class LLMS_Metabox_Textarea_W_Tags_Field extends LLMS_Metabox_Field implements M
 	 *
+	 * @since Unknown
+	 *
 	 * @param array $_field Array containing information about field
```
```suggestion
	 * @param array $_field Array containing information about field.
```

eri-trabiccolo commented on includes/traits/llms-trait-singleton.php:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Singleton class trait.
+ *
+ * @package  LifterLMS_REST/Traits
```
```suggestion
 * @package LifterLMS/Traits
```

eri-trabiccolo commented on includes/traits/llms-trait-singleton.php:
```diff
@@ -0,0 +1,44 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LLMS_REST_Trait_Singleton class..
```
```suggestion
 * LLMS_REST_Trait_Singleton class
```

eri-trabiccolo commented on includes/traits/llms-trait-singleton.php:
```diff
@@ -0,0 +1,44 @@
+	 *
+	 * @since [version]
+	 *
+	 * @return obj instance of the LifterLMS_REST_API class.
```
```suggestion
	 * @return object Instance of the class.
```

eri-trabiccolo committed:
Fix unhandled error when scraping a certificate to export producing a PHP fatal

eri-trabiccolo committed:
Fix documentation for the get_export() method return type

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

Commit:
Merge pull request #1277 from eri-trabiccolo/certificate-export-handle-errors

Fix unhandled error when scraping a certificate to export producing a fatal

Commit:
Merge pull request #1271 from gocodebox/jquery-migrate

Remove usage of deprecated jQuery.on( 'ready', ... )

thomasplevy committed:
new utilities

thomasplevy committed:
add sample data imports for testing

thomasplevy committed:
test restrictions

thomasplevy committed:
Merge branch 'dev' of github.com:gocodebox/lifterlms into dev

thomasplevy committed:
pkg build: 2.2.0

thomasplevy committed:
build: 4.3.1

thomasplevy committed:
Merge branch 'trunk' of github.com:gocodebox/lifterlms into rel/4.3.1

thomasplevy committed:
build: 4.3.1

thomasplevy committed:
fix typos

Commit:
Merge pull request #1279 from gocodebox/rel/4.3.1

Build 4.3.1

Commit:
Merge pull request #1273 from gocodebox/dev

v4.3.1: Dev to Trunk

thomasplevy committed:
exclude 'protected' post types from wp 5.5. xml sitemap

Commit:
Merge pull request #1282 from gocodebox/sitepmaps

exclude 'protected' post types from wp 5.5. xml sitemap

thomasplevy committed:
Merge branch 'trunk' of github.com:gocodebox/lifterlms into dev

thomasplevy committed:
Publish

 - @lifterlms/llms-e2e-test-utils@2.1.1
 - @lifterlms/scripts@1.2.2

thomasplevy committed:
update webpack config generator to have a configurable source path and default source files to './src' instead of './assets/src'

thomasplevy committed:
build: 4.3.2

Commit:
Merge pull request #1284 from gocodebox/build/4.3.2

build: 4.3.2

thomasplevy committed:
bump tested version to 5.5

Commit:
Merge pull request #1283 from gocodebox/dev

v4.3.2: Dev to Trunk

thomasplevy committed:
Publish

 - @lifterlms/llms-e2e-test-utils@2.1.2
 - @lifterlms/scripts@1.2.3

thomasplevy committed:
resolve files

thomasplevy committed:
fix script file refereneces

thomasplevy committed:
update lerna

thomasplevy committed:
Publish

 - @lifterlms/scripts@1.2.4

thomasplevy committed:
fix jest config to work when installed outside the core monorepo

thomasplevy committed:
Publish

 - @lifterlms/scripts@1.3.0

thomasplevy committed:
dont use imports

thomasplevy committed:
Publish

 - @lifterlms/scripts@1.3.1

thomasplevy committed:
Publish

 - @lifterlms/scripts@1.3.2

eri-trabiccolo committed:
Doc: Fix inline comment invalid end char for a bunch of files

eri-trabiccolo committed:
Apply suggestions from code review

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

Commit:
Merge pull request #1287 from eri-trabiccolo/invalid-end-char

Doc: Fix inline comment invalid end char for a bunch of files

thomasplevy committed:
reduce logical complexity of reporting related asset registration and enqueue and ensure analytics script is not enqueued on all settings screens

thomasplevy committed:
fix cs

thomasplevy committed:
add WP 5.4 to WP Version Matrix

thomasplevy committed:
put legends on the top of the chart automatically

Commit:
Merge pull request #1290 from gocodebox/issue/1269

put legends on the top of the chart automatically

Commit:
Merge pull request #1289 from gocodebox/issue/1288

Refactor asset enqueue/registration logic for reporting/analytics related scripts

thomasplevy committed:
asset handler

thomasplevy committed:
fix css

thomasplevy committed:
stop registering the main llms script in the ajax class, use asset handler

thomasplevy committed:
stop reference wp_ajax_data for ajax nonce information

thomasplevy committed:
fix cs

thomasplevy committed:
Merge branch 'asset-handler' of github.com:gocodebox/lifterlms into fix-core-script-double-registration

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

Commit:
Merge pull request #1293 from gocodebox/fix-core-script-double-registration

Fix core script double registration

thomasplevy committed:
rename 'file_slug' to 'file_name' & make optional, defaulting to the asset handle

thomasplevy committed:
use LLMS_Assets for script and style definitions

thomasplevy committed:
Merge branch 'asset-cleanup' of github.com:gocodebox/lifterlms into asset-handler

thomasplevy committed:
use empty strings in favor of null for post_content and post_excerpt post fields

thomasplevy committed:
update docs

thomasplevy committed:
ignore db warnings

thomasplevy committed:
strict comparator

thomasplevy committed:
fix docs

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

thomasplevy committed:
fix return

Commit:
Merge pull request #1295 from gocodebox/generator

Fix issue with empty post content and except via the LLMS_Generator class

thomasplevy committed:
Merge branch 'trunk' of github.com:gocodebox/lifterlms into dev

thomasplevy committed:
build: 4.3.3

Commit:
Merge pull request #1296 from gocodebox/build/4.3.3

build: 4.3.3

Commit:
Merge pull request #1291 from gocodebox/dev

v4.3.3: Dev to Trunk

thomasplevy committed:
make asset handler more extendable

thomasplevy committed:
merge dev

thomasplevy committed:
move inline script functions into LLMS_Assets

thomasplevy committed:
fix cs

thomasplevy committed:
doc fix

thomasplevy committed:
use core methods for deprecated warnings and errors

thomasplevy committed:
version doc

thomasplevy committed:
add tests for llms_deprecated_function()

eri-trabiccolo committed:
Fix wrong next/previous lesson retrieved in presence of unpublished lessons

thomasplevy committed:
dont use deprecated method

thomasplevy committed:
Merge branch 'asset-handler' of github.com:gocodebox/lifterlms into deprecated-function

thomasplevy committed:
dont log to llms-log

Commit:
Merge pull request #1298 from gocodebox/deprecated-function

Removes "duplicate" logic from `llms_deprecated_function()` in favor of core method

thomasplevy committed:
dont use deprecated method

eri-trabiccolo committed:
Fix previous lesson for the following scenario:
If the first lesson of a section is drafted, I would expect the previous
lesson of the second lesson in that section to be the last lesson of the
previous section but it's actually returning false (as in there's no previous lesson).

eri-trabiccolo committed:
Fix unpublished lessons in course's syllabus still counted when displaying lesssons order.

thomasplevy committed:
update contributor-faces, add limit to contributors generator, update contributors list

thomasplevy committed:
add build command & update docs

thomasplevy committed:
remove unused deps and scripts

thomasplevy committed:
Merge branch 'issue/1294' of github.com:eri-trabiccolo/lifterlms into dev

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

Commit:
Merge pull request #1292 from gocodebox/asset-handler

asset handler class and tests

thomasplevy committed:
reduce code complexity and add additional parameters to filters

thomasplevy committed:
add tests for inline asset methods

thomasplevy committed:
fix cs

thomasplevy committed:
minor code improvements, update docs, add tests

thomasplevy committed:
update docs

thomasplevy committed:
fix cs

thomasplevy committed:
update docs

thomasplevy committed:
remove unused private method

thomasplevy committed:
strict comparator, prevent object unserialization

thomasplevy committed:
update cs to latest

thomasplevy committed:
remvoe double periods

thomasplevy committed:
fix docs

Commit:
Merge pull request #1306 from gocodebox/update-question-docs

Updates docs and makes minor code improvements to the LLMS_Question model

Commit:
Create codeql-analysis.yml

Commit:
remove release branch

Commit:
remove build steps

Commit:
Create ossar-analysis.yml

Commit:
Create shiftleft-analysis.yml

Commit:
Update shiftleft-analysis.yml

Commit:
Update codeql-analysis.yml

thomasplevy committed:
make travis configs importable and extendable

thomasplevy committed:
Merge branch 'trunk' of github.com:gocodebox/lifterlms into dev

thomasplevy committed:
fix undefined variable, per #1195

Commit:
Merge pull request #1307 from gocodebox/issue/1195

fix undefined variable

thomasplevy committed:
add deprecation note

thomasplevy committed:
build: 4.4.0

thomasplevy committed:
build: 4.4.0

Commit:
Merge pull request #1308 from gocodebox/rel/4.4.0

Build: 4.4.0

thomasplevy committed:
merge dev

thomasplevy commented:
I accidentally merged this into dev and had to rebase. Code review has been applied and reopening PR.
